{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2623bc4a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.165241Z",
     "iopub.status.busy": "2025-03-08T15:06:24.164381Z",
     "iopub.status.idle": "2025-03-08T15:06:24.176195Z",
     "shell.execute_reply": "2025-03-08T15:06:24.174944Z"
    },
    "papermill": {
     "duration": 0.021454,
     "end_time": "2025-03-08T15:06:24.178846",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.157392",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "def f1(x):\n",
    "    if (x > 0):\n",
    "        x = 3*x \n",
    "        x = x / 2\n",
    "    return x\n",
    "\n",
    "def f2(x):\n",
    "    if (x > 0):\n",
    "        x = 3*x\n",
    "    x = x / 2\n",
    "    return x\n",
    "\n",
    "print(f1(0))\n",
    "print(f2(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ecdc9abf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.189365Z",
     "iopub.status.busy": "2025-03-08T15:06:24.188935Z",
     "iopub.status.idle": "2025-03-08T15:06:24.195306Z",
     "shell.execute_reply": "2025-03-08T15:06:24.193986Z"
    },
    "papermill": {
     "duration": 0.014304,
     "end_time": "2025-03-08T15:06:24.197648",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.183344",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "2\n",
      "1\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "def function1(length):\n",
    "    if length > 0:\n",
    "        print(length)\n",
    "        function1(length - 1)\n",
    "        \n",
    "print(function1(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62d0f9ee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.208242Z",
     "iopub.status.busy": "2025-03-08T15:06:24.207411Z",
     "iopub.status.idle": "2025-03-08T15:06:24.214437Z",
     "shell.execute_reply": "2025-03-08T15:06:24.213110Z"
    },
    "papermill": {
     "duration": 0.015065,
     "end_time": "2025-03-08T15:06:24.216975",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.201910",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "def compute(n,x,y) :\n",
    "    if n==0 : return x\n",
    "    return compute(n-1,x+y,y)\n",
    "\n",
    "print(compute(1,2,3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c9df802c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.228133Z",
     "iopub.status.busy": "2025-03-08T15:06:24.227191Z",
     "iopub.status.idle": "2025-03-08T15:06:24.235412Z",
     "shell.execute_reply": "2025-03-08T15:06:24.234021Z"
    },
    "papermill": {
     "duration": 0.016527,
     "end_time": "2025-03-08T15:06:24.237894",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.221367",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "def valid_dna4(dna):\n",
    "    for c in dna:\n",
    "        if not c in 'acgtACGT':\n",
    "            return False\n",
    "    return True\n",
    "    \n",
    "def valid_dna3(dna):\n",
    "    for c in dna:\n",
    "        flag = c in 'acgtACGT'\n",
    "    return flag\n",
    "\n",
    "def valid_dna1(dna):\n",
    "    for c in dna:\n",
    "        if c in 'acgtACGT':\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "            \n",
    "print(valid_dna1('atgcatkjghlhg'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bf320f9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.249025Z",
     "iopub.status.busy": "2025-03-08T15:06:24.248635Z",
     "iopub.status.idle": "2025-03-08T15:06:24.260179Z",
     "shell.execute_reply": "2025-03-08T15:06:24.258832Z"
    },
    "papermill": {
     "duration": 0.019938,
     "end_time": "2025-03-08T15:06:24.262713",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.242775",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "3\n",
      "3\n",
      "3\n",
      "3\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def count1(dna, base):\n",
    "    i = 0\n",
    "    for c in dna:\n",
    "        if c == base:\n",
    "            i += 1 \n",
    "    return i\n",
    "\n",
    "def count2(dna, base):\n",
    "    i = 0 \n",
    "    for j in range(len(dna)):\n",
    "        if dna[j] == base:\n",
    "            i += 1 \n",
    "    return i \n",
    "\n",
    "def count3(dna, base):\n",
    "    match = [c == base for c in dna]\n",
    "    return sum(match)\n",
    "\n",
    "def count4(dna, base):\n",
    "    return dna.count(base)\n",
    "\n",
    "def count5(dna, base):\n",
    "    return len([i for i in range(len(dna)) if dna[i] == base])\n",
    "\n",
    "def count6(dna,base):\n",
    "    return sum(c == base for c in dna)\n",
    "\n",
    "\n",
    "dna = 'aaatttgggccc'\n",
    "base = 'a'\n",
    "\n",
    "print(count1(dna, base))\n",
    "print(count2(dna, base))\n",
    "print(count3(dna, base))\n",
    "print(count4(dna, base))\n",
    "print(count5(dna, base))\n",
    "print(count6(dna, base))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f430fcb0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.274131Z",
     "iopub.status.busy": "2025-03-08T15:06:24.273032Z",
     "iopub.status.idle": "2025-03-08T15:06:24.298927Z",
     "shell.execute_reply": "2025-03-08T15:06:24.297631Z"
    },
    "papermill": {
     "duration": 0.036054,
     "end_time": "2025-03-08T15:06:24.303261",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.267207",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed time:  0.0006370000000002207\n",
      "Elapsed time:  0.0008300000000005525\n",
      "Elapsed time:  0.000868000000000535\n",
      "Elapsed time:  0.0001299999999995194\n",
      "Elapsed time:  0.000834999999999475\n",
      "Elapsed time:  0.000942999999999472\n",
      "tgaccagtacacccctgcattatatgattgaaattccccatgagttatatgccggtcgtctaccaggaccgcaggtgcgaccgacacggcccgtctactactaagtgttccacttccatcctaagctaggtcgttctcaaacggatatcgccgcagctgaccgatctgcgacgacattttgagtacatcggataaaagttaaccagctaaaaggtaataaccgacccagttcaaggaggcgttctaagtacccccctctgatatgatggattgttacttttctaatttgagacctcgacgatgcgactgccatgtcaggtgtcagttccgacgttatgcgtacaacgatcggcggtctggagagttcagcagggccgatcgtcgccgcaaaatagcgtcagaaggtctgtttatgtttgtagctgaaagttgcgaaagtcgtgtagaagagagacgcatcgcaaagacaatccccagacactcaggtagaaacttaggccacccagtgtctcctatcacgcatattgaattaccctttttacgaggttttgaattctatattactaattccccacatttaccacaagtcgtaccctcctaaggccggcgacgcctttgttacgagtagtgttatgtgtaaggtctattgcaatagcttgagaatagaatccaggaccgatccccgatacctctgctgccgcaggccagatgctatgctgccgaggcccttgtctcgtctttttgcttatgccccgatagttgctacgcagaagtattacagctacgtcacctggtttgtcagcgtgtatgattgcagtcttgactaatccggcctcaaaccgccgttggtcctttttccgcaaataggcgtacccataaaagaggatgcgcgccacactgagaggaatttggtgctctgaccatttttggctctttcgatactctaattcatcctgcagaccacggcggaattatgactttatttagcatctgtttctagctgattcatcaaggcgctagagagttcagcctgtgatagcctctaacttcgtctaggagttggcgccaaacgaatctcttgaatgggccggaagtatacaagcggtcgggtaactctcgagtatgcaccttaattctcacaaagttttaggagtaaacccccttggggaccccacgggatccggggagggtctgtgcggaattcgcgtcagacactggattatgacgaatggtagtggattgccgaatctaacggctcggccccgctcgacggtgagcttgcatcttcgacggacgctgcgcaggaggcctctagcatttaagacacggctccaactgtatttgcgtcaagttgattaactaacacggtaagagacgtagaaacggacgctctccgccgcaactaggaacccttgtcctgaggtgactcgcaagctatttcccttcctctggttaacagttggatgcaatgtactgtaatagggagaccgggcgctcctaagctcgtggtggcatagcgcctggcgggtgtgtgagatgagcacagtcatccccagtacgtcctcgccaaaggtttaccatctgtcatcattcttctacacccccgtatgagtcagtttcccaatttgaaaacggtctggattgttcggctacatgccgttcagccaattccacacaagaggcagcttggaaacctagggaaaacaagcgccaccccgttcttcaccacttgaccgcgatccggagataataggtcgtaacgacgttcgggaccgtgatctttctgaccaaccctggccacaaagtgaatgctgaatttccctagcacatacctgaatagtggtgaccgggatcttttttgccatcaacctatagacttggtgaactatgttagatcatatatgccagcaccgtgccacgtgaaaaccttatgctgtgttccacgcaatcagagtgtgctcaccagaggcagcggtgggcttttacgtagcggagattattatgaggaaattagtggactaatgtcgtggtgtcaccgtgttgctcactgaagtgaagtcgccgagagtgccttctggagcacgttttcccggcccactgagattctgattcgaaactggcctactatcccaaggctccacgccaatagttgatctggcgataatcccatatgtagtcagggactgacacacaatgtccttgaaccacaccttacctttttctcacaagccattttggtctgatcatgcagacagtaacgtggtaacgggtccatgatagtgcccatttgtatggtgagccccgctcatcgtgtccttgttaggaagttaggcaggcccgaatatctcttccctccgaatgcgagcttcctcagaacttctgaccaaagggtcttctcctctccagccaaaaagacttggtagggctgacccttgagccgcgcagataccgtcctagatgtggcggtaacggcatgcaagaaggatcccatggcattcggacgcgtacgtacctcatcatgctgtgctactaagccctaccctgccaagacattaaagcgtggatgttctaggtcttactttcaattttgactattagagaggagcaatgtgcccgtatgttgtcaaaaccccacacttcgagaactaaacagtgcctaacgcgagcaaccaaatttgctaagtgtgctctaagcacggaccctaacagttcgttgcgcccatatctgggggaccgggtactggtgcggaatcggcatgatggacttagcatctccgtcttacgttacgggtgtaatagatatcttgtaccaacgcacgatgcggcgtatcattagagagctttcggtgtccaatgcgaaagaggtagacttgctgtttacccagccagtaaaacgggcatagcgcactgtccagcggggtcaggctggtgggccatcttagctacactcctggggcataatccaccggtccctcctcttggcttctgccaacctatctcttatcccatctgttcgaccatcgagagcaataaccctgactcccattaaaaccatgtaatagccgcctttcagcaaacgcttcgttccatttattggcttattttgccccgaagggcgacagagggtatctaaaaccgccatgccgcactcgcgagcgaaatcctcccctgaactagcgccatctcgtcaacatagcatccagtaaaccccagcgcccggcctaccgtatttgtctctgtatgcactacttgttttcgtcatagttactagacggctactccgaccagactactaggtgaggtgcgggatacggcaatcgatatagttaggatgccgggacgtttggtttttctaatccttacggtatgtagtgtgacaactgctcaccactctcatatcggagattggaatagaaattattaacgcttattcgggaaggaacgcgccccgcgaatcagggattggcagacaagcacagatggcggggcagtgctaagtaacgacttgatcagaccacggaactacaaccccgttagtggcatctggtttggtctgctgaatacctacgcgctaagaacgaagcgcatccaagtcctaagctagggcaattaatagacatttttttgttgtcttcaatcatggtgcacctttacgggatcctggattacaaaggacctaacacagtccaaggagaatcattcatgtcagcaagacgaaatagagcaaactttggcttgtttcggatatcggccgtagtaaactgtcttaatgacacctcgctagatcgccgattgtcggggaccaagaactttagtagcttcgcgggactctccgccagtcctgatgggactagcgatacatgtaaagctaataacacacatcgatgtaccttcgattcgtagtacaggccaaatgtccagtctgtttagtatgctgtaaatcggattttttgatagcgataggtgaatagggtgcctggagtaccaagggacctagcactcgcgggcgttctgatacgcagacttcatgcgaagcttagatgatgaacgggactttataatttcagtcccagggtatgcccttgtaaccgccgcagattccttgggcccgcaaacggccacgtagttggtagtgcaatggccggcagaagcgttcatcggttttgtgagcgcaatagcgttgtatttcggggacaatgcctaaggcgaactcatgacataccgaatccgtggattttgagacgactcgtactagttctccgtttgatctttatgaatgctagccgggtatgggcgccatataatgcaatcgtcccgacatacagttggcggagtaaggtttctcatgctgaacccgcgtccgcactctgagagtgaccctacttgtcggcaccattgtcctgtacgtgtgctccgaactacttatcttgagacccagagctagggtaactgccaatattgcaacgcggtatcagtactcaacctaaagactacttatatttaggacgtatttcctatgtatagggttagtgctgatgaactctgtcgcattcctataccgctcccgccgttcgacctaacagcgtaacgcctcgcttggctcagttatttcatttccggcgaaactcttatcccgacatcgcgattgcacaatctagcggcgagccatctgtttgaagggcgagtgactctttacccatgcagtaaacttctgatttgaagagccggtcgtgcctagccggctgttgattcccgactactaataagaacccccccatggctacgaccgatgaagaaggtcacgctgcatgggaattatgcacggagcggaaatggtaatggcacccggtgcaagcacatgtgggaagggaattgcaatcgcaaagggactgaaatgagactcacttggatgttgttccacgctcaatgagtcgttaacgcattataagataggccaaacacagggacggtcccgggtcacagaccattgagcataaaaccggtgactccttgtatcctgcagtggatgagggcctgcttaaaggccagaacaattttgcagagtccagttttttttgtgaagctttctgatttgacacaacgaatttccctccagaggtggcatattcgcaccagactttgacccctcttgcctgagtaatcgactgtaactgtgtttgtcgaatgccccgatcgtggcttacggtatacgggaaccatgccacggcccacacattgctttactacccaatcgttggcccacgtccgtgtgctttagacgctcttggaatcgtgggcgtcgcggcaacagattagggatagttctgacggccattaagaaactgcgggctgatgtacggctcccctgcacaacagcatcacatgggctgcgcctagtcctaatgacgttcccgtaatgtttcagtggactgctagatcgctgaacttcaatggcctaatgggtcgaattcgcatcgtcgttcgatcattggtccgtgatttaccctaagtaacgacattggttagggacttaaatcattagggtaaactcccgtggcgttttggagaatgggcctccgaggattgtttgatgctaccgtgttaggtcaaatatcatcataaagtgacttaaattggtttgatcgggcaagtcgtgtagtgccgttcgtatgttgagaggtcaactccaaactaaactcgggagagttttacatatatgaccgtgttcttactggcataaagggctggggatgtcgtgtctaagctggagttacagcgcgctgccaaggcttaattacgctttctacgaggatgctaggtggaatcttacactgggagttggtcggcaagagaacgtagtcggcgaccctaaaggtgaacctgaaaaggaagagctccgagtactagtgtcgcgacgtaacaggtacgccggccatgaacgtcaccagcgagcttagccgttgccgaggtagagctcaggaagactgtgggcatctagtgtttggagctcgtgtagtaatctaggcgcataaaagacaaatttccttacgatgccctacatgacctcttggctctaggaacctggggccgcctaattagagagagccgggaatctcgagcttaatggcttgttcacggcgagtcctaaagcttgcgtctgtttctgtgagcgggccacaagagtctctcactcgcatataccctcggccgtaagcccgcgctggctaagtaagaacattgacttctaaccgagccccgaagcaactctaactgcgaagctccactatgtacaggtttgcctagtcgaatctcgacaggtgtcatcaaaacaacgtcgttttctttgtcgtctacttcctttcctaagggtacccaattgtgacccagtgtgcctgtaggctagctaatagaggtcccccacttaactcatgagtaaccgaaggaggcgagactgcgctcgtagcgatacacgagtgctaaagtatctatatgatttgtgctgactccctcctcacataatggggcgggtgtacgcaggcgcgcaagccctcctgcgcgtacaagcccttgagttaatgtctatacgatgacttgtaagaatgtaccaacatctggggtccatgaccccgtgaggagcgaccccttggttgatagcggtggctttaggattccggagtgcaagacctcttacacagtgtttcgcgttacacctgctattatccaaagaaaatcgcttgcttgcttccggtgtaatatattccccggtcacttaagcagcttaacctaaccacaaccactgacatataaggtctgagtcatatccatcttacctattcttgataaccgcgtttagtattggagaagcgggtcgtcaagccttcctaccgatcgcaagttatgcatcaagtactagcgtttttgtgtacgcggtgcttccgacaaacagccctttggtaaccataccactacctagccggaattcactcgtcttagcacgagcaattctgacccgggttccgcctcggtgagggtacggtcgcggttcagcgcacacacgtataatttacacgcgcccagcgtcgcgggccgtatgcctatgctcttcctatatatgcagggtggtccggcgcttgattttgtctatgaacattctaaaaggagcaggaaggataatgggttctacggacttatccggacccgccacgaacgtagcgaacaagccccgcgcattgctatactcccactcaacgactccttgaacgtccgggtcgctttctttacaacttgtaagctggacttgtcagcttaacactccaaatcttgttctggagactaacacgtggcgtctagcaagccccgatcgtcgaaccgtgctatccttacagtctcttcctatccggtgaagcccttacggcatcgcgggccagcgcgtcattatgggggtaagcatcaaagcccgacgttactcaactatattgctatagtaaccgacccactctggcgagtttcctggtacgtaatttttagcacggaaatggaagggtgatcgtcgccccgcgagcgtgctgctaaaatagacaaaacgagcccaaacaaatatgcgtctaagaagttcttcatcttgtaccattataccgcaggcaaggcaagtagtagacacgggacaagaggcgatatcgaccccccgtcataccggcaacgctgcggcagggttaggaactaatccaagttaatttgatcctgacttaagaccttctccgtacagcatctgacactatcacaatacctcaagaaatatcggaagtgaccgccgcacctagggatgtgggctccagtccgctcaataaaagcgggccgaagagtggtagtgaaaaggagcaaaataccagcgcccacttgtgctgacaatcctggccctcgtcacaccatcaagtctcacccgaggtgggatcgctcgacccaagttcacgtagagggacgttcccctggttagcaggcccaaccatcaagaagctttttatactttcgacttgacttttatggcttttgtcctgtttcagatcatatgacacaaaacatttacggtatatgtcgaggtgctcgataccaatctcttacggttatccgggctctaatatggagtaacttgacccaaaatatgtccacggcgcgatcagagttattggtagttgcttcgagatgcgagcagttgacaagtcttaatattacttaagacgcaagtaagcaagaggtaacggtgcaacctgagagtaccgccatgttgcaattctgtcccgccgtaagagtacatcctaccaacctctgtaaacctctcaggttggcgtggttagccctatagcataagttcgacacgcgtaaggtatattgtccgtagaatctattatagcagtttacgctacgtgtctcgcagtgccgaatgttaccgaatagaaggaagcgtagcgacgcgtcggttggaggagagttcgatcggcggcaatcctgtagtatcactagagtctgacaggaaatgctttgttttagtatctgactcgatgctcgtacgcagtctaaagcgagagccacgcagaccagcacacaatttaccgtcaattctccttccgagaggcgtagttcgaacaatcacttacaccctaaggaataacgccgaaaattggaatctcccgctaaagggacgcaggccccgccttcgtgtgctagccgatcccagaatatgactattcaaggggggtgctcactggtaccttaaaagccagcgaaatacaccgccatcggaaccccttgaaccctcttgctagcgccggcttttgcagcactaagcccaatgcagagtcctgaggcagtacgagagggtccttatgtagtatacttcgactcagagatattataccgtgcccgcttaactctcaacttaaaataggtccacgccataaatcccagcagcggcgtatgcgggttgtaaagtcctataagcgcgaccagcggtgcttatccccccgaattattcagtcaggaagattacactttaccactgaaggctttattctggtgcgccgtgatctatattaccgagcttggaaattgcgaaaacagactttagtgacaagcccatgacccgtgtacaattcacggcctgccacgtgattcgatagatctcattgtagtccttgttcagcgaagggggcggcccgttatcatgtaactgcgcagcccggttgggatggacaaagtttagctgagactcctatgagggacgggccgcacacgcataacctgatcgcattaacctatggacataaaccgatacccccaacgacccgtgttgcgtcctcccactcctctagaacccattcggctgtggtcgcgggcatcgaatgtggtggtgtaggatcagcgagtacctagtgcggtctgcttgcaagtcactgcgatgaagaggatattaaacttgatgccatcgtatttccgctgagaatcactctagtccattgcaccggaaagcggagcaaccattctattttaggcacaaagcacagtgttcccaaccatctataccggactaccgaactcgccctaaatcctcatcgacttctagatatgtgcaccaggtcaacagagttagacgtgtgattattatatctaatggtacggaccggagcgaaccggtgtttcggaccaagaagtccgggaactgtgcgatctagctctagagtctgtgagattctcggaacaggcggtttagtcgattacaagcttctatccgtattcgtgtcgtggcgttaaggtgagcatggtttactatcgcacaacgcagtctctacggaagccgccaagagtctgcccagccgttgagggccaagcttgcgttcctatgggggaactttggagagtcacatacgggttaggcatgacgcgggttgaaaccttggaacgatactccgatacatttggccccaacccataactcctaccgaggtac\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:8: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "  \n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:10: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:15: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "  from ipykernel import kernelapp as app\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:17: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "  app.launch_new_instance()\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:22: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:24: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:30: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:32: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:37: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:39: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:44: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n",
      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:46: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import random\n",
    "import time\n",
    "def create_dna(n, alphabet='acgt'):\n",
    "    return ''.join([random.choice(alphabet) for i in range(n)])\n",
    "dna = create_dna(10000)\n",
    "base = 'c'\n",
    "\n",
    "start_time = time.clock()\n",
    "count1(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "start_time = time.clock()\n",
    "count2(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "start_time = time.clock()\n",
    "count3(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "\n",
    "start_time = time.clock()\n",
    "count4(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "start_time = time.clock()\n",
    "count5(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "start_time = time.clock()\n",
    "count6(dna, base)\n",
    "end_time = time.clock()\n",
    "\n",
    "elapsed_time = end_time - start_time\n",
    "print(\"Elapsed time: \", elapsed_time)\n",
    "\n",
    "\n",
    "print(dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad3c9bd2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.315022Z",
     "iopub.status.busy": "2025-03-08T15:06:24.314149Z",
     "iopub.status.idle": "2025-03-08T15:06:24.320127Z",
     "shell.execute_reply": "2025-03-08T15:06:24.318638Z"
    },
    "papermill": {
     "duration": 0.014718,
     "end_time": "2025-03-08T15:06:24.322819",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.308101",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Function to read a FASTA file\n",
    "\n",
    "# try:\n",
    "#     f = open(\"myfile.fa\")\n",
    "# except IOError:\n",
    "#     print(\"File myfile.fa does not exist!\")\n",
    "    \n",
    "# seqs = {}\n",
    "# for line in f: # reads every line entry  in the file\n",
    "#     line = line.rstrip() # removes any newline at the end\n",
    "#     if line[0] == '>': # every line starts with the > identifier so we check to distinguish header from sequence\n",
    "#         words = line.split() # tokenise the strings into separate  words\n",
    "#         words = words[0][1:] # ignore the > identifier in the first string to select the name of the sequence\n",
    "#         seqs[name] = '' # i.e. if we're in a header line, create a new sequence with the name\n",
    "#     else:\n",
    "#         seqs[name] = seqs[name]+  line #  else append the dna sequence in the current line to the dict\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3e00f15c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.334493Z",
     "iopub.status.busy": "2025-03-08T15:06:24.333681Z",
     "iopub.status.idle": "2025-03-08T15:06:24.339321Z",
     "shell.execute_reply": "2025-03-08T15:06:24.338039Z"
    },
    "papermill": {
     "duration": 0.01438,
     "end_time": "2025-03-08T15:06:24.341947",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.327567",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Script  for building a dictionary containing all sequences from a FASTA file\n",
    "\n",
    "\n",
    "# try:\n",
    "#     f = open(\"myfile.fa\")\n",
    "# except IOError:\n",
    "#     print(\"File myfile.fa does not exist!\")\n",
    "    \n",
    "# seqs = {}\n",
    "# for line in f: # reads every line entry  in the file\n",
    "#     line = line.rstrip() # removes any newline at the end\n",
    "#     if line[0] == '>': # every line starts with the > identifier so we check to distinguish header from sequence\n",
    "#         words = line.split() # tokenise the strings into separate  words\n",
    "#         words = words[0][1:] # ignore the > identifier in the first string to select the name of the sequence\n",
    "#         seqs[name] = '' # i.e. if we're in a header line, create a new sequence with the name\n",
    "#     else:\n",
    "#         seqs[name] = seqs[name]+  line #  else append the dna sequence in the current line to the dict\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e124476",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.354380Z",
     "iopub.status.busy": "2025-03-08T15:06:24.353924Z",
     "iopub.status.idle": "2025-03-08T15:06:24.360893Z",
     "shell.execute_reply": "2025-03-08T15:06:24.359607Z"
    },
    "papermill": {
     "duration": 0.016457,
     "end_time": "2025-03-08T15:06:24.363540",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.347083",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id1 AAGGGTTCCAGGTCC\n",
      "id2 AAGGGTTCCAGGTCC\n"
     ]
    }
   ],
   "source": [
    "dict = {'id1': \"AAGGGTTCCAGGTCC\", 'id2': \"AAGGGTTCCAGGTCC\"}\n",
    "\n",
    "for name, seq in dict.items():\n",
    "    print(name, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e25e2cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.375228Z",
     "iopub.status.busy": "2025-03-08T15:06:24.374414Z",
     "iopub.status.idle": "2025-03-08T15:06:24.380514Z",
     "shell.execute_reply": "2025-03-08T15:06:24.379259Z"
    },
    "papermill": {
     "duration": 0.014673,
     "end_time": "2025-03-08T15:06:24.383071",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.368398",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py', '-f', '/tmp/tmpigyhdxyu.json', '--HistoryManager.hist_file=:memory:']\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.argv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f63e0a5f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.395236Z",
     "iopub.status.busy": "2025-03-08T15:06:24.394145Z",
     "iopub.status.idle": "2025-03-08T15:06:24.400078Z",
     "shell.execute_reply": "2025-03-08T15:06:24.398935Z"
    },
    "papermill": {
     "duration": 0.014538,
     "end_time": "2025-03-08T15:06:24.402599",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.388061",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import sys\n",
    "# import getopt\n",
    "# def usage():\n",
    "#     print \"\"\"\n",
    "#     processfasta.py : reads a FASTA file and builds a dictionary with all sequences\n",
    "#     bigger than a given length\n",
    "    \n",
    "#     processfasta.py [-h] [-l <length>] <filename> \n",
    "    \n",
    "#     -h              print this message\n",
    "    \n",
    "#     -l <length>     filter all sequences with a length smaller than <length> \n",
    "#                     (default <length> = 0)\n",
    "    \n",
    "#     <filename>      the file has to be in FASTA format\n",
    "    \n",
    "#     \"\"\"\n",
    "# # the [] around an arg like '-l <length>' indicates it's an optional argument\n",
    "    \n",
    "# o, a = getopt.getopt(sys.argv[1:], 'l:h') # first arg - ignore the name of program; 2nd arg - list all args the user can provide with a dash (-) command; ':' after a parameter indicates that the opttion expects a value\n",
    "# # o - list of optional argument\n",
    "# # a - list of required arguments\n",
    "\n",
    "# opts = {}\n",
    "# seqLen = 0;\n",
    "\n",
    "# for k,v in o:\n",
    "#     opts[k]= v\n",
    "# if '-h' in opts.keys():\n",
    "#     usage(); sys.exit() # quits the program completely\n",
    "# if len(a) < 1: # check to see if the file name is available\n",
    "#     usageg(); sys.exit(\"input fasta file is missing\")\n",
    "# if '-l' in opts.keys():\n",
    "#     if int(opts['l']) < 0:\n",
    "#         print(\"Length of sequence should be positive!\"); sys.exit(0)\n",
    "#     seqLen = opts['-l']\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a2595e45",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.414052Z",
     "iopub.status.busy": "2025-03-08T15:06:24.413681Z",
     "iopub.status.idle": "2025-03-08T15:06:24.420974Z",
     "shell.execute_reply": "2025-03-08T15:06:24.419554Z"
    },
    "papermill": {
     "duration": 0.016088,
     "end_time": "2025-03-08T15:06:24.423595",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.407507",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "def get_extension1(filename):\n",
    "    return(filename.split(\".\")[-1])\n",
    "\n",
    "def get_extension2(filename):\n",
    "    import os.path\n",
    "    return(os.path.splitext(filename)[1])\n",
    "\n",
    "def get_extension3(filename):\n",
    "    return filename[filename.rfind('.'):][1:]\n",
    "\n",
    "\n",
    "print(get_extension2(\"myfile\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "508df69f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.435752Z",
     "iopub.status.busy": "2025-03-08T15:06:24.434925Z",
     "iopub.status.idle": "2025-03-08T15:06:24.444250Z",
     "shell.execute_reply": "2025-03-08T15:06:24.442980Z"
    },
    "papermill": {
     "duration": 0.018118,
     "end_time": "2025-03-08T15:06:24.446769",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.428651",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.81\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "print(Bio.__version__)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2fdcd441",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.459058Z",
     "iopub.status.busy": "2025-03-08T15:06:24.458103Z",
     "iopub.status.idle": "2025-03-08T15:06:24.615419Z",
     "shell.execute_reply": "2025-03-08T15:06:24.614313Z"
    },
    "papermill": {
     "duration": 0.170328,
     "end_time": "2025-03-08T15:06:24.621984",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.451656",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reverse complement is AACCGGGTCAGT\n",
      "Help on class Seq in module Bio.Seq:\n",
      "\n",
      "class Seq(_SeqAbstractBaseClass)\n",
      " |  Seq(data, length=None)\n",
      " |  \n",
      " |  Read-only sequence object (essentially a string with biological methods).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to protein sequences).\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Seq\n",
      " |      _SeqAbstractBaseClass\n",
      " |      abc.ABC\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash of the sequence as a string for comparison.\n",
      " |      \n",
      " |      See Seq object comparison documentation (method ``__eq__`` in\n",
      " |      particular) as this has changed in Biopython 1.65. Older versions\n",
      " |      would hash on object identity.\n",
      " |  \n",
      " |  __init__(self, data, length=None)\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - data - Sequence, required (string)\n",
      " |       - length - Sequence length, used only if data is None or a dictionary (integer)\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, you can also create a Seq object directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      \n",
      " |      To create a Seq object with for a sequence of known length but\n",
      " |      unknown sequence contents, use None for the data argument and pass\n",
      " |      the sequence length for the length argument. Trying to access the\n",
      " |      sequence contents of a Seq object created in this way will raise\n",
      " |      an UndefinedSequenceError:\n",
      " |      \n",
      " |      >>> my_undefined_sequence = Seq(None, 20)\n",
      " |      >>> my_undefined_sequence\n",
      " |      Seq(None, length=20)\n",
      " |      >>> len(my_undefined_sequence)\n",
      " |      20\n",
      " |      >>> print(my_undefined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
      " |      \n",
      " |      If the sequence contents is known for parts of the sequence only, use\n",
      " |      a dictionary for the data argument to pass the known sequence segments:\n",
      " |      \n",
      " |      >>> my_partially_defined_sequence = Seq({3: \"ACGT\"}, 10)\n",
      " |      >>> my_partially_defined_sequence\n",
      " |      Seq({3: 'ACGT'}, length=10)\n",
      " |      >>> len(my_partially_defined_sequence)\n",
      " |      10\n",
      " |      >>> print(my_partially_defined_sequence)\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined\n",
      " |      >>> my_partially_defined_sequence[3:7]\n",
      " |      Seq('ACGT')\n",
      " |      >>> print(my_partially_defined_sequence[3:7])\n",
      " |      ACGT\n",
      " |  \n",
      " |  ungap(self, gap='-')\n",
      " |      Return a copy of the sequence without the gap character(s) (DEPRECATED).\n",
      " |      \n",
      " |      The gap character now defaults to the minus sign, and can only\n",
      " |      be specified via the method argument. This is no longer possible\n",
      " |      via the sequence's alphabet (as was possible up to Biopython 1.77):\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\")\n",
      " |      >>> my_dna\n",
      " |      Seq('-ATA--TGAAAT-TTGAAAA')\n",
      " |      >>> my_dna.ungap(\"-\")\n",
      " |      Seq('ATATGAAATTTGAAAA')\n",
      " |      \n",
      " |      This method is DEPRECATED; please use my_dna.replace(gap, \"\") instead.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add a sequence or string to this sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq(\"MELKI\") + \"LV\"\n",
      " |      Seq('MELKILV')\n",
      " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
      " |      MutableSeq('MELKILV')\n",
      " |  \n",
      " |  __bytes__(self)\n",
      " |  \n",
      " |  __contains__(self, item)\n",
      " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> MutableSeq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string.\n",
      " |      \n",
      " |      Sequences are equal to each other if their sequence contents is\n",
      " |      identical:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == mutable_seq\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      \n",
      " |      Note that the sequence objects themselves are not identical to each\n",
      " |      other:\n",
      " |      \n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> seq1 is seq2\n",
      " |      False\n",
      " |      \n",
      " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
      " |      objects:\n",
      " |      \n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == b\"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == bytearray(b\"ACGT\")\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence as a single letter or as a sequence object.\n",
      " |      \n",
      " |      If the index is an integer, a single letter is returned as a Python\n",
      " |      string:\n",
      " |      \n",
      " |      >>> seq = Seq('ACTCGACGTCG')\n",
      " |      >>> seq[5]\n",
      " |      'A'\n",
      " |      \n",
      " |      Otherwise, a new sequence object of the same class is returned:\n",
      " |      \n",
      " |      >>> seq[5:8]\n",
      " |      Seq('ACG')\n",
      " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
      " |      >>> mutable_seq[5:8]\n",
      " |      MutableSeq('ACG')\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply the sequence object by other and assign.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      \n",
      " |      Note that this is different from in-place multiplication. The ``seq``\n",
      " |      variable is reassigned to the multiplication result, but any variable\n",
      " |      pointing to ``seq`` will remain unchanged:\n",
      " |      \n",
      " |      >>> seq = Seq('ATG')\n",
      " |      >>> seq2 = seq\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      True\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG')\n",
      " |      >>> seq2\n",
      " |      Seq('ATG')\n",
      " |      >>> id(seq) == id(seq2)\n",
      " |      False\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Return an iterable of the sequence.\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply sequence by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> MutableSeq('ATG') * 2\n",
      " |      MutableSeq('ATGATG')\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence string on the left.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> \"LV\" + Seq(\"MELKI\")\n",
      " |      Seq('LVMELKI')\n",
      " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
      " |      MutableSeq('LVMELKI')\n",
      " |      \n",
      " |      Adding two sequence objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by sequence.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string.\n",
      " |  \n",
      " |  back_transcribe(self, inplace=False)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence.back_transcribe()\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      >>> sequence.back_transcribe(inplace=True)\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
      " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
      " |      back-transcribe method will ensure any U becomes T.\n",
      " |      \n",
      " |      Trying to back-transcribe a protein sequence will replace any U for\n",
      " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Seq('MAIVMGRT')\n",
      " |  \n",
      " |  complement(self, inplace=None)\n",
      " |      Return the complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement(inplace=False)\n",
      " |      Seq('GCTAA')\n",
      " |      \n",
      " |      In contrast, ``complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement(inplace=False)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement(inplace=True)\n",
      " |      MutableSeq('GCT')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  complement_rna(self, inplace=False)\n",
      " |      Return the complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement_rna()\n",
      " |      Seq('GCU')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").complement_rna()\n",
      " |      Seq('GCUAA')\n",
      " |      \n",
      " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").complement()\n",
      " |      Seq('GCT')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.complement_rna()\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.complement_rna(inplace=True)\n",
      " |      MutableSeq('GCU')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('GCU')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  count(self, sub, start=None, end=None)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      The number of occurrences of substring argument sub in the\n",
      " |      (sub)sequence given by [start:end] is returned as an integer.\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
      " |      objects, like that of Python strings, do a non-overlapping search, this\n",
      " |      may not give the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      For an overlapping search, use the ``count_overlap`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |  \n",
      " |  count_overlap(self, sub, start=None, end=None)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      For a non-overlapping search, use the ``count`` method:\n",
      " |      \n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
      " |      the ``count`` method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  endswith(self, suffix, start=None, end=None)\n",
      " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.find(\"AUG\", 4)\n",
      " |      15\n",
      " |  \n",
      " |  index(self, sub, start=None, end=None)\n",
      " |      Return the lowest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the lowest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Raises a ValueError if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.index(\"AUG\")\n",
      " |      3\n",
      " |      \n",
      " |      The next typical start codon can then be found by starting the search\n",
      " |      at position 4:\n",
      " |      \n",
      " |      >>> my_rna.index(\"AUG\", 4)\n",
      " |      15\n",
      " |      \n",
      " |      This method performs the same search as the ``find`` method.  However,\n",
      " |      if the subsequence is not found, ``find`` returns -1 which ``index``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  islower(self)\n",
      " |      Return True if all ASCII characters in data are lowercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  isupper(self)\n",
      " |      Return True if all ASCII characters in data are uppercase.\n",
      " |      \n",
      " |      If there are no cased characters, the method returns False.\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
      " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
      " |      objects. These arguments will be concatenated with the calling sequence\n",
      " |      as the spacer:\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Joining the letters of a single sequence:\n",
      " |      \n",
      " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
      " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
      " |  \n",
      " |  lower(self, inplace=False)\n",
      " |      Return the sequence in lower case.\n",
      " |      \n",
      " |      An lower-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``upper`` method.\n",
      " |  \n",
      " |  lstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.lstrip()\n",
      " |      Seq('ACGT ')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the leading end instead.  The order of the characters to be removed\n",
      " |      is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
      " |      Seq('TTACG')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.lstrip(inplace=False)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.lstrip(inplace=True)\n",
      " |      MutableSeq('ACGT ')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT ')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  replace(self, old, new, inplace=False)\n",
      " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
      " |      \n",
      " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
      " |      >>> s\n",
      " |      Seq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      Seq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
      " |      >>> m\n",
      " |      MutableSeq('ACGTAACCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
      " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
      " |      >>> m\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      >>> t\n",
      " |      MutableSeq('XYZGTAXYZCGGTT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement(self, inplace=None)\n",
      " |      Return the reverse complement as a DNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement(inplace=False)\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      Any U in the sequence is treated as a T:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement(inplace=False)\n",
      " |      Seq('AATCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement_rna`` returns an RNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement(inplace=False)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement(inplace=True)\n",
      " |      MutableSeq('TCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('TCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  reverse_complement_rna(self, inplace=False)\n",
      " |      Return the reverse complement as an RNA sequence.\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
      " |      Seq('UCG')\n",
      " |      \n",
      " |      Any T in the sequence is treated as a U:\n",
      " |      \n",
      " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
      " |      Seq('AAUCG')\n",
      " |      \n",
      " |      In contrast, ``reverse_complement`` returns a DNA sequence:\n",
      " |      \n",
      " |      >>> Seq(\"CGA\").reverse_complement(inplace=False)\n",
      " |      Seq('TCG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"CGA\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      >>> my_seq.reverse_complement_rna()\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('CGA')\n",
      " |      \n",
      " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
      " |      MutableSeq('UCG')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('UCG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
      " |      ``inplace=True``.\n",
      " |  \n",
      " |  rfind(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
      " |      3\n",
      " |  \n",
      " |  rindex(self, sub, start=None, end=None)\n",
      " |      Return the highest index in the sequence where subsequence sub is found.\n",
      " |      \n",
      " |      With optional arguments start and end, return the highest index in the\n",
      " |      sequence such that the subsequence sub is contained within the sequence\n",
      " |      region [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq or MutableSeq object to search for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rindex(\"AUG\")\n",
      " |      15\n",
      " |      \n",
      " |      The location of the typical start codon before that can be found by\n",
      " |      ending the search at position 15:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
      " |      3\n",
      " |      \n",
      " |      This method performs the same search as the ``rfind`` method.  However,\n",
      " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
      " |      raises a ValueError:\n",
      " |      \n",
      " |      >>> my_rna.rindex(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: ...\n",
      " |      >>> my_rna.rfind(\"T\")\n",
      " |      -1\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences by splitting the sequence from the right.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``rsplit`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |      \n",
      " |      See also the split method, which splits the sequence starting from the\n",
      " |      beginning:\n",
      " |      \n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |  \n",
      " |  rstrip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.rstrip()\n",
      " |      Seq(' ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      from the trailing end instead.  The order of the characters to be\n",
      " |      removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
      " |      Seq('ACGACGTT')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.rstrip(inplace=False)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.rstrip(inplace=True)\n",
      " |      MutableSeq(' ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
      " |      \n",
      " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      For consistency with the ``split`` method of Python strings, any\n",
      " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
      " |      default value\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L')\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR')\n",
      " |      Seq('L')\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR')\n",
      " |      Seq('KGAR*L')\n",
      " |      \n",
      " |      See also the rsplit method, which splits the sequence starting from the\n",
      " |      end:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR')\n",
      " |      Seq('L')\n",
      " |  \n",
      " |  startswith(self, prefix, start=None, end=None)\n",
      " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None, inplace=False)\n",
      " |      Return a sequence object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      With default arguments, leading and trailing whitespace is removed:\n",
      " |      \n",
      " |      >>> seq = Seq(\" ACGT \")\n",
      " |      >>> seq.strip()\n",
      " |      Seq('ACGT')\n",
      " |      >>> seq\n",
      " |      Seq(' ACGT ')\n",
      " |      \n",
      " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
      " |      instead.  The order of the characters to be removed is not important:\n",
      " |      \n",
      " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
      " |      Seq('')\n",
      " |      \n",
      " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
      " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
      " |      in-place and returned.\n",
      " |      \n",
      " |      >>> seq = MutableSeq(\" ACGT \")\n",
      " |      >>> seq.strip(inplace=False)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq(' ACGT ')\n",
      " |      >>> seq.strip(inplace=True)\n",
      " |      MutableSeq('ACGT')\n",
      " |      >>> seq\n",
      " |      MutableSeq('ACGT')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
      " |      is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  transcribe(self, inplace=False)\n",
      " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      >>> sequence.transcribe()\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
      " |      \n",
      " |      >>> sequence.transcribe(inplace=True)\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      >>> sequence\n",
      " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      Trying to transcribe an RNA sequence has no effect.\n",
      " |      If you have a nucleotide sequence which might be DNA or RNA\n",
      " |      (or even a mixture), calling the transcribe method will ensure\n",
      " |      any T becomes U.\n",
      " |      \n",
      " |      Trying to transcribe a protein sequence will replace any\n",
      " |      T for Threonine with U for Selenocysteine, which has no\n",
      " |      biologically plausible rational.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Seq('MAIVMGRU')\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences. It should not\n",
      " |      be used on protein sequences as any result will be biologically\n",
      " |      meaningless.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         Defaults to the minus sign.\n",
      " |      \n",
      " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
      " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
      " |      pn a ``MutableSeq`` object.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*')\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@')\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR')\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*')\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR')\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR')\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR')\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR')\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead\n",
      " |  \n",
      " |  upper(self, inplace=False)\n",
      " |      Return the sequence in upper case.\n",
      " |      \n",
      " |      An upper-case copy of the sequence is returned if inplace is False,\n",
      " |      the default value:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq, MutableSeq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*')\n",
      " |      \n",
      " |      The sequence is modified in-place and returned if inplace is True:\n",
      " |      \n",
      " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      >>> my_seq.lower()\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper()\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPeeK*')\n",
      " |      \n",
      " |      >>> my_seq.lower(inplace=True)\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('vhltpeek*')\n",
      " |      >>> my_seq.upper(inplace=True)\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      >>> my_seq\n",
      " |      MutableSeq('VHLTPEEK*')\n",
      " |      \n",
      " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
      " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
      " |      \n",
      " |      See also the ``lower`` method.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  defined\n",
      " |      Return True if the sequence is defined, False if undefined or partially defined.\n",
      " |      \n",
      " |      Zero-length sequences are always considered to be defined.\n",
      " |  \n",
      " |  defined_ranges\n",
      " |      Return a tuple of the ranges where the sequence contents is defined.\n",
      " |      \n",
      " |      The return value has the format ((start1, end1), (start2, end2), ...).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from _SeqAbstractBaseClass:\n",
      " |  \n",
      " |  __array_ufunc__ = None\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/Bio/pairwise2.py:283: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  BiopythonDeprecationWarning,\n"
     ]
    }
   ],
   "source": [
    "# Problem: Find out from what species an unknown DNA sequence came from\n",
    "\n",
    "# myseq.fa\n",
    "# >sequence_unknown\n",
    "# CATGCTACGGGTGCTAAAAGCATTACGCCCTATAGTGATTTTCGAGACATACTGTGTTTTTAAATATAGTATTGCC\n",
    "\n",
    "# fasta_string = open(\"myseq.fa\").read() # to open the file and read the seq into a var\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO, SearchIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# fasta_string = Seq(\"CATGCTACGGGTGCTAAAAGCATTACGCCCTATAGTGATTTTCGAGACATACTGTGTTTTTAAATATAGTATTGCC\")\n",
    "# result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", fasta_string) \n",
    "#blastn - program thatt searches nt against nt; nt - database (nonredundant database containign everything in NCBI)\n",
    "\n",
    "# from Bio.Blast import NCBIXML\n",
    "# blast_record = NCBIXML.read(result_handle)\n",
    "# len(blast_record.alignments)\n",
    "\n",
    "# E_VALUE_THRESH = 0.1\n",
    "# for alignment in blast_record.alignments:\n",
    "#     for hsp in alignmentt.hsps:\n",
    "#         if hsp.expect < E_VALUE_THRESH:\n",
    "#             print('***Alignment***')\n",
    "#             print('sequence: ', alignment.title)\n",
    "#             print('length: ', alignment.length)\n",
    "#             printt('e value: ', hsp.expect)\n",
    "#             print(hsp.query)\n",
    "#             print(hsp.match)\n",
    "#             print(hsp.sbjct)\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "myseq = Seq(\"ACTGACCCGGTT\")\n",
    "print('reverse complement is ' + str(myseq.reverse_complement()))\n",
    "help(Seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1227829d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.638205Z",
     "iopub.status.busy": "2025-03-08T15:06:24.637788Z",
     "iopub.status.idle": "2025-03-08T15:06:24.644025Z",
     "shell.execute_reply": "2025-03-08T15:06:24.642893Z"
    },
    "papermill": {
     "duration": 0.017078,
     "end_time": "2025-03-08T15:06:24.647271",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.630193",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WASYLSYIPCSYGGAMFYVNPRSKDIIPKSYN*DLDMVLFCLSYRRLCSFLLFFGLLLILLWHLR*\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/Bio/Seq.py:2808: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "my_seq = Seq(\"TGGGCCTCATATTTATCCTATATACCATGTTCGTATGGTGGCGCGATGTTCTACGTGAATCCACGTTCGAAGGACATCATACCAAAGTCGTACAATTAGGACCTCGATATGGTTTTATTCTGTTTATCGTATCGGAGGTTATGTTCTTTTTTGCTCTTTTTCGGGCTTCTTCTCATTCTTCTTTGGCACCTACGGTAGAG\")\n",
    "\n",
    "print(my_seq.translate())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "74b46c09",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-08T15:06:24.662845Z",
     "iopub.status.busy": "2025-03-08T15:06:24.661672Z",
     "iopub.status.idle": "2025-03-08T15:06:24.666989Z",
     "shell.execute_reply": "2025-03-08T15:06:24.665667Z"
    },
    "papermill": {
     "duration": 0.015677,
     "end_time": "2025-03-08T15:06:24.669396",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.653719",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# fasta_string = \">sequence_name\\nMTEITAAMVKELRESTGAGMMDCKNALSETNGDFDKAVQLLREKGLGKAAKKADRLAAEG\\nLVSVKVSDDFTIAAMRPSYLSYEDLDMTFVENEYKALVAELEKENEERRRLKDPNKPEHK\\nIPQFASRKQLSDAILKEAEEKIKEELKAQGKPEKIWDNIIPGKMNSFIADNSQLDSKLTL\\nMGQFYVMDDKKTVEQVIAEKEKEFGGKIKIVEFICFEVGEGLEKKTEDFAAEVAAQL\\n\"\n",
    "# result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", fasta_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49df908",
   "metadata": {
    "papermill": {
     "duration": 0.006075,
     "end_time": "2025-03-08T15:06:24.681883",
     "exception": false,
     "start_time": "2025-03-08T15:06:24.675808",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kaggle": {
   "accelerator": "none",
   "dataSources": [],
   "dockerImageVersionId": 30445,
   "isGpuEnabled": false,
   "isInternetEnabled": false,
   "language": "python",
   "sourceType": "notebook"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 12.604497,
   "end_time": "2025-03-08T15:06:25.311278",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2025-03-08T15:06:12.706781",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
